### PR TITLE
TEST/UCP/RMA: Refactor RMA and atomic tests, add memtype support

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

File: test/gtest/ucp/test_ucp_atomic.h [hunk 1]
```diff
@@ -11,120 +11,5 @@
 #include "test_ucp_memheap.h"
 
 
-class test_ucp_atomic : public test_ucp_memheap {
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_atomic.h:14]: this file is not needed anymore, can be removed

File: test/gtest/ucp/test_ucp_rma.cc [hunk 2]
```diff
@@ -19,290 +23,143 @@ class test_ucp_rma : public test_ucp_memheap {
...
 
-UCS_TEST_P(test_ucp_rma, nonblocking_put_nbi_flush_worker) {
-    test_blocking_xfer(static_cast<nonblocking_send_func_t>(&test_ucp_rma::nonblocking_put_nbi),
-                       DEFAULT_SIZE, DEFAULT_ITERS,
-                       1, false, false);
-    test_blocking_xfer(static_cast<nonblocking_send_func_t>(&test_ucp_rma::nonblocking_put_nbi),
-                       DEFAULT_SIZE, DEFAULT_ITERS,
-                       1, true, false);
-}
+        for (size_t i = 0; i < pairs.size(); ++i) {
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_rma.cc:78]: should it be two-dimensional `for `?

File: test/gtest/ucp/test_ucp_rma.cc [hunk 3]
```diff
@@ -19,290 +23,143 @@ class test_ucp_rma : public test_ucp_memheap {
...
+               void *expected_data, void *arg) {
+        ucs_status_ptr_t status_ptr;
+        ucp_request_param_t param;
 
-        ucs::fill_random(memheap_addr, ucs_min(max_size, 16384U));
-        status = ucp_get_nbi(e->ep(), (void *)&expected_data[0], expected_data.length(),
-                             (uintptr_t)memheap_addr, rkey);
-        ASSERT_UCS_OK_OR_INPROGRESS(status);
+        param.op_attr_mask = 0;
+        status_ptr = ucp_get_nbx(sender().ep(), expected_data, size,
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_rma.cc:57]: there is common code between `get_b()` and `get_nbi()`, can be separate function `do_get()` as for PUT operation

File: test/gtest/ucp/test_ucp_atomic.cc [hunk 4]
```diff
@@ -5,229 +5,173 @@
...
-
-    prev    = *(T*)memheap_addr;
-    val     = (T)ucs::rand() * (T)ucs::rand();
-
-    amo_req = test_ucp_atomic::ucp_atomic_fetch<T>(e->ep(), FOP,
-                                                   val, &result, memheap_addr, rkey);
-    if(UCS_PTR_IS_PTR(amo_req)){
-        request_wait(amo_req);
+protected:
+    static const uint64_t POST_ATOMIC_OPS  = UCS_BIT(UCP_ATOMIC_OP_ADD) |
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_atomic.cc:77]: align by `|`?

File: test/gtest/ucp/test_ucp_atomic.cc [hunk 5]
```diff
@@ -5,229 +5,173 @@
...
+                                             UCS_BIT(UCP_ATOMIC_OP_AND) |
+                                             UCS_BIT(UCP_ATOMIC_OP_OR) |
+                                             UCS_BIT(UCP_ATOMIC_OP_XOR);
+    static const uint64_t FETCH_ATOMIC_OPS = POST_ATOMIC_OPS |
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_atomic.cc:81]: align by `|`?

File: test/gtest/ucp/test_ucp_atomic.cc [hunk 6]
```diff
@@ -5,229 +5,173 @@
...
-    void *amo_req;
-
-    prev = *(T*)memheap_addr;
-    if ((ucs::rand() % 2) == 0) {
-        compare = prev; /* success mode */
-    } else {
-        compare = ~prev; /* fail mode */
+private:
+    static T atomic_op_result(ucp_atomic_op_t op, T x, T y, T z) {
+        /* coverity[switch_selector_expr_is_constant] */
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_atomic.cc:107]: do we need this below?
or can this suppression be removed here?

File: test/gtest/ucp/test_ucp_atomic.cc [hunk 7]
```diff
@@ -5,229 +5,173 @@
...
+        case UCP_ATOMIC_OP_XOR:
+            return "XOR";
+        default:
+            return 0;
+        }
     }
 
-    EXPECT_EQ(prev, result);
+    ucs_status_t do_atomic(ucp_atomic_op_t op, size_t size, void *target_ptr,
+                           ucp_rkey_h rkey, T value, ucp_request_param_t& param) {
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_atomic.cc:146]: `ucp_request_param_t &param`

File: test/gtest/ucp/test_ucp_memheap.h [hunk 8]
```diff
@@ -14,51 +14,31 @@
...
     /*
-     * @param [in]  max_size       Maximal size of data to send.
-     * @param [in]  memheap_addr   VA to perform the RMA operation to,
-     * @param [in]  rkey           Memheap remote key.
-     * @param [out] expected_data  What should the memheap contain at the given
-     *                             address after the operation (also can be used
-     *                             as a source/destination data).
+     * Function type for memheap send operation
+     *
+     * @param [in]  size            Size of data to send
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_memheap.h:19]: trailing `.` for all sentences?

File: test/gtest/ucp/test_ucp_memheap.cc [hunk 9]
```diff
@@ -7,248 +7,97 @@
...
+    ucs_assert(!(mem_map_flags & (UCP_MEM_MAP_ALLOCATE | UCP_MEM_MAP_FIXED)));
 
-    sender().connect(&receiver(), get_ep_params());
+    mem_buffer memheap(num_iters * size + alignment, target_mem_type);
+    padding = UCS_PTR_BYTE_DIFF(memheap.ptr(),
+                                ucs_align_up_pow2_ptr(memheap.ptr(), alignment));
+    ucs_assert(padding >= 0);
+    ucs_assert(padding < alignment);
 
+    /* Allocate heap */
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_memheap.cc:41]: should we start all comments with lowercase letters?

File: test/gtest/ucp/test_ucp_memheap.h [hunk 10]
```diff
@@ -14,51 +14,31 @@
...
+                                      ucp_rkey_h rkey, void *expected_data,
+                                      void *arg);
 
 protected:
-    const static size_t DEFAULT_SIZE  = 0;
-    const static int    DEFAULT_ITERS = 0;
-
-    void test_blocking_xfer(blocking_send_func_t send, size_t len, int max_iters,
-                            size_t alignment, bool malloc_allocate, bool is_ep_flush);
+    virtual void init();
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_memheap.h:36]: virtual is excessive, or was it added just to make it more informative?

File: test/gtest/ucp/test_ucp_rma.cc [hunk 11]
```diff
@@ -19,290 +23,143 @@ class test_ucp_rma : public test_ucp_memheap {
...
 
-UCS_TEST_P(test_ucp_rma, nonblocking_stream_get_nbi_flush_worker) {
-    test_nonblocking_implicit_stream_xfer(static_cast<nonblocking_send_func_t>(&test_ucp_rma::nonblocking_get_nbi),
-                       DEFAULT_SIZE, DEFAULT_ITERS,
-                       1, false, false);
-    test_nonblocking_implicit_stream_xfer(static_cast<nonblocking_send_func_t>(&test_ucp_rma::nonblocking_get_nbi),
-                       DEFAULT_SIZE, DEFAULT_ITERS,
-                       1, true, false);
-}
+            ucs_memory_type_t mem_types[2] = {send_mem_type, target_mem_type};
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_rma.cc:138]: 1. the size of the array can be removed
2. do we need whitespace after/before `{`/`}`?

File: test/gtest/ucp/test_ucp_atomic.cc [hunk 12]
```diff
@@ -5,229 +5,173 @@
...
-
-void test_ucp_atomic::init() {
-    const char *atomic_mode =
-                    (GetParam().variant == UCP_ATOMIC_MODE_CPU)    ? "cpu" :
-                    (GetParam().variant == UCP_ATOMIC_MODE_DEVICE) ? "device" :
-                    (GetParam().variant == UCP_ATOMIC_MODE_GUESS)  ? "guess" :
-                    "";
-    modify_config("ATOMIC_MODE", atomic_mode);
-    test_ucp_memheap::init();
+#include <ucp/core/ucp_types.h> // for atomic mode
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_atomic.cc:10]: use `/**/` or change the comment in
https://github.com/openucx/ucx/blob/18616602fee2e5f3d26af20b462915a9b53fbc64/test/gtest/ucp/test_ucp_rma.cc#L12

File: test/gtest/ucp/test_ucp_atomic.cc [hunk 13]
```diff
@@ -5,229 +5,173 @@
...
-        flush_worker(*e);
-    } else {
-        ASSERT_UCS_OK(status);
+    void misaligned_post(size_t size, void *target_ptr, ucp_rkey_h rkey,
+                         void *expected_data, void *arg)
+    {
+        *(T*)expected_data = *(T*)target_ptr; /* remote should not change */
+
+        ucp_request_param_t param;
+        param.op_attr_mask = 0;
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_atomic.cc:49]: align by `=`?

File: test/gtest/ucp/test_ucp_atomic.cc [hunk 14]
```diff
@@ -5,229 +5,173 @@
...
-        scoped_log_handler slh(hide_errors_logger);
-        status = ucp_atomic_add64(e->ep(), 0, (uintptr_t)memheap_addr + 1, rkey);
+        ucp_atomic_op_t op = *(ucp_atomic_op_t*)arg;
+        T value            = (T)ucs::rand() * (T)ucs::rand();
+        T prev             = *(T*)target_ptr;
+        *(T*)expected_data = atomic_op_result(op, value, prev, 0);
+
+        ucp_request_param_t param;
+        param.op_attr_mask = 0;
+        do_atomic(op, size, target_ptr, rkey, value, param);
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_atomic.cc:40]: check `ucs_status_t` returned from `do_atomic`?

File: test/gtest/ucp/test_ucp_atomic.cc [hunk 15]
```diff
@@ -5,229 +5,173 @@
...
-    ASSERT_UCS_OK(test_ucp_atomic::ucp_atomic_post_nbx<T>(e->ep(), OP, val,
-                                                          memheap_addr, rkey));
+        ucp_request_param_t param;
+        param.op_attr_mask = UCP_OP_ATTR_FIELD_REPLY_BUFFER;
+        param.reply_buffer = &reply_data;
 
-    expected_data.resize(sizeof(T));
-    *(T*)&expected_data[0] = nbx_atomic_op_val<T, OP>(val, prev);
-}
+        do_atomic(op, size, target_ptr, rkey, value, param);
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_atomic.cc:71]: check `ucs_status_t` returned from `do_atomic`?

File: test/gtest/ucp/test_ucp_atomic.cc [hunk 16]
```diff
@@ -5,229 +5,173 @@
...
-{
-    T val, prev;
 
-    prev   = *(T*)memheap_addr;
-    val    = (T)ucs::rand() * (T)ucs::rand();
+    void fetch(size_t size, void *target_ptr, ucp_rkey_h rkey,
+               void *expected_data, void *arg)
+    {
+        ucp_atomic_op_t op = *(ucp_atomic_op_t*)arg;
+        T value             = (T)ucs::rand() * (T)ucs::rand();
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_atomic.cc:60]: unaligned by `=`?

File: test/gtest/ucp/test_ucp_rma.cc (hunk 2 above)
Comment (yosefe) [reply test/gtest/ucp/test_ucp_rma.cc:78]: i wanted to reuse the function ucs::supported_mem_type_pairs(), and this way code is somewhat shorter

File: test/gtest/ucp/test_ucp_memheap.cc (hunk 9 above)
Comment (yosefe) [reply test/gtest/ucp/test_ucp_memheap.cc:41]: IMO better start with uppercase

File: test/gtest/ucp/test_ucp_memheap.h (hunk 10 above)
Comment (yosefe) [reply test/gtest/ucp/test_ucp_memheap.h:36]: yes, to emphasize it overrides a virtual function

File: test/gtest/ucp/test_ucp_rma.cc (hunk 11 above)
Comment (yosefe) [reply test/gtest/ucp/test_ucp_rma.cc:138]: 1. will remove
2. no, it's same as `(` and `[`

**Comment by yosefe** [comment]
failure is https://github.com/openucx/ucx/issues/5558
bot:pipe:retest

**Comment by yosefe** [comment]
@dmitrygx can you pls look again?

**Pull request merged**